src/Cmfd.cpp

```diff
@@ -606,311 +539,140 @@ FP_PRECISION Cmfd::computeKeff(int moc_iteration) {
-                   mat[cell][g*(_num_cmfd_groups+4)+g+2];
-
-          /* Bottom surface */
-          if (y != 0)
-            val -= _SOR_factor * vec_x[row - _num_x * _num_cmfd_groups] *
-                   mat[cell][g*(_num_cmfd_groups+4)+1] /
-                   mat[cell][g*(_num_cmfd_groups+4)+g+2];
-
-          /* Group-to-group */
-          for (int e = 0; e < _num_cmfd_groups; e++) {
-            if (e != g)
-              val -= _SOR_factor * vec_x[cell*_num_cmfd_groups+e] *
-                     mat[cell][g*(_num_cmfd_groups+4)+2+e] /
-                     mat[cell][g*(_num_cmfd_groups+4)+g+2];
-          }
-
-          /* Right surface */
-          if (x != _num_x - 1)
-            val -= _SOR_factor * vec_x[row + _num_cmfd_groups] *
-                   mat[cell][g*(_num_cmfd_groups+4)+_num_cmfd_groups+2] /
-                   mat[cell][g*(_num_cmfd_groups+4)+g+2];
-
-          /* Top surface */
-          if (y != _num_y - 1)
-            val -= _SOR_factor * vec_x[row + _num_cmfd_groups*_num_x] *
-                   mat[cell][g*(_num_cmfd_groups+4)+_num_cmfd_groups+3] /
-                   mat[cell][g*(_num_cmfd_groups+4)+g+2];
-
-          vec_x[row] = val;
-        }
-      }
-    }
-
-    /* Compute the average residual */
-    residual = 0.0;
-    for (int i = 0; i < _num_x*_num_y*_num_cmfd_groups; i++) {
-      if (vec_x[i] != 0.0)
-        residual += pow((vec_x[i] - _flux_temp[i]) / vec_x[i], 2);
-    }
-    residual = pow(residual, 0.5) / (_num_x*_num_y*_num_cmfd_groups);
-
-    /* Increment the interations counter */
-    iter++;
-
-    log_printf(DEBUG, "GS iter: %d, res: %f", iter, residual);
-
-    if (residual < conv && iter > 10)
-      break;
-  }
-
-  log_printf(DEBUG, "linear solver iterations: %d", iter);
-}
-
-
-/**
  * @brief Rescale the initial and converged flux arrays.
+ * @details The diffusion problem is a generalized eigenvalue problem and
+ *         therefore the solution is independent of flux level. This method
+ *         rescales the input flux and converged flux to both have an average
+ *         fission source of 1.0 in each group in each cell.
  */
 void Cmfd::rescaleFlux() {
 
-  FP_PRECISION sum_new, sum_old, scale_val;
-
   /* Rescale the new and old flux to have an avg source of 1.0 */
-  matrix_multiplication(_M, _new_flux, _new_source, _num_x*_num_y, 
-      _num_cmfd_groups);
-  sum_new = pairwise_sum(_new_source, _num_x*_num_y*_num_cmfd_groups);
-  scale_val = _num_x*_num_y*_num_cmfd_groups / sum_new;
-  vector_scale(_new_flux, scale_val, _num_x*_num_y*_num_cmfd_groups);
-  matrix_multiplication(_M, _old_flux, _old_source, _num_x*_num_y, 
-          _num_cmfd_groups);
-  sum_old = pairwise_sum(_old_source, _num_x*_num_y*_num_cmfd_groups);
-  scale_val = _num_x*_num_y*_num_cmfd_groups / sum_old;
-  vector_scale(_old_flux, scale_val, _num_x*_num_y*_num_cmfd_groups);
+  matrixMultiplication(_M, _new_flux, _new_source);
+  matrixMultiplication(_M, _old_flux, _old_source);
```

Comment:
I tend to agree that this would be a nice feature too. But I would also be
fine without this for this PR if it's not how @samuelshaner wants to spend
his time.
On Jul 27, 2015 12:11 PM, "Geoffrey Gunow" <notifications@github.com> wrote:

> In src/Cmfd.cpp
> <https://github.com/mit-crpg/OpenMOC/pull/184#discussion_r35574215>:
>
> >    /* Rescale the new and old flux to have an avg source of 1.0 */
> > -  matrix_multiplication(_M, _new_flux, _new_source, _num_x*_num_y,
> > -      _num_cmfd_groups);
> > -  sum_new = pairwise_sum(_new_source, _num_x*_num_y*_num_cmfd_groups);
> > -  scale_val = _num_x*_num_y*_num_cmfd_groups / sum_new;
> > -  vector_scale(_new_flux, scale_val, _num_x*_num_y*_num_cmfd_groups);
> > -  matrix_multiplication(_M, _old_flux, _old_source, _num_x*_num_y,
> > -          _num_cmfd_groups);
> > -  sum_old = pairwise_sum(_old_source, _num_x*_num_y*_num_cmfd_groups);
> > -  scale_val = _num_x*_num_y*_num_cmfd_groups / sum_old;
> > -  vector_scale(_old_flux, scale_val, _num_x*_num_y*_num_cmfd_groups);
> > +  matrixMultiplication(_M, _new_flux, _new_source);
> > +  matrixMultiplication(_M, _old_flux, _old_source);
>
> We might want to consider making a matrix class that supports operator
> overloading for matrix multiplication
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mit-crpg/OpenMOC/pull/184/files#r35574215>.
>